avogadro/rendering/solidpipeline.cpp

```diff
@@ -0,0 +1,132 @@
+/******************************************************************************
+  This source file is part of the Avogadro project.
+  This source code is released under the 3-Clause BSD License, (see "LICENSE").
+******************************************************************************/
+
+#include "solidpipeline.h"
+
+#include "solid_vs.h"
+
+#include "solid_first_fs.h"
+
+#include <iostream>
+
+namespace Avogadro::Rendering {
+
+SolidPipeline::SolidPipeline()
+{
+}
+
+void initializeFramebuffer(GLuint *outFBO, GLuint *texRGB, GLuint *texDepth)
+{
+  glGenFramebuffers(1, outFBO);
+  glBindFramebuffer(GL_FRAMEBUFFER, *outFBO);
+
+  glGenTextures(1, texRGB);
+  glBindTexture(GL_TEXTURE_2D, *texRGB);
+  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, *texRGB, 0);
+
+  glGenTextures(1, texDepth);
+  glBindTexture(GL_TEXTURE_2D, *texDepth);
+  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, *texDepth, 0);
+}
+
+void SolidPipeline::initialize()
+{
+  initializeFramebuffer(&m_renderFBO, &m_renderTexture, &m_depthTexture);
+  resize(1024, 768);
+
+  glGenBuffers(1, &m_screenVBO);
+  glBindBuffer(GL_ARRAY_BUFFER, m_screenVBO);
+  glBufferData(GL_ARRAY_BUFFER, sizeof(m_fullscreenQuad), m_fullscreenQuad, GL_STATIC_DRAW);
+
+  m_screenVertexShader.setType(Shader::Vertex);
+  m_screenVertexShader.setSource(solid_vs);
+  if (!m_screenVertexShader.compile())
+    std::cout << m_screenVertexShader.error() << std::endl;
+
+  m_firstFragmentShader.setType(Shader::Fragment);
+  m_firstFragmentShader.setSource(solid_first_fs);
+  if (!m_firstFragmentShader.compile())
+    std::cout << m_firstFragmentShader.error() << std::endl;
+
+  m_firstStageShaders.attachShader(m_screenVertexShader);
+  m_firstStageShaders.attachShader(m_firstFragmentShader);
+  if (!m_firstStageShaders.link())
+    std::cout << m_firstStageShaders.error() << std::endl;
+}
+
+void SolidPipeline::begin()
+{
+  glGetIntegerv(GL_FRAMEBUFFER_BINDING, (GLint *) &m_defaultFBO);
+  glBindFramebuffer(GL_FRAMEBUFFER, m_renderFBO);
+  GLenum drawBuffersList[1] = {GL_COLOR_ATTACHMENT0};
+  glDrawBuffers(1, drawBuffersList);
+
+  GLfloat tmp[5];
+  glGetFloatv(GL_COLOR_CLEAR_VALUE, tmp);
+  glGetFloatv(GL_DEPTH_CLEAR_VALUE, tmp + 4);
+  glClearColor(0.0, 0.0, 0.0, 0.0);
+  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+  glClearColor(tmp[0], tmp[1], tmp[2], tmp[3]);
+  glClearDepthf(tmp[4]);
```

Comment:
This line crashes on Mac - it looks like it's de-referencing a null pointer. If it's commented out, it otherwise seems to run. (This is perhaps because we're not using the 3.2 core profile.)

Reply:
Yep, sorry, `glClearDepthf` is GL 4.1+ only, should use `glClearDepth` instead.